src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/ProxyWhitelist.java

```diff
@@ -24,274 +24,111 @@
 
 package org.jenkinsci.plugins.scriptsecurity.sandbox.whitelists;
 
+import edu.umd.cs.findbugs.annotations.NonNull;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Field;
 import java.lang.reflect.Method;
-
-import edu.umd.cs.findbugs.annotations.NonNull;
-import net.jcip.annotations.GuardedBy;
-import org.jenkinsci.plugins.scriptsecurity.sandbox.Whitelist;
-import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.WeakHashMap;
-import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.jenkinsci.plugins.scriptsecurity.sandbox.Whitelist;
 
 /**
  * Aggregates several whitelists.
  */
 public class ProxyWhitelist extends Whitelist {
-    @GuardedBy("lock")
-    private Collection<? extends Whitelist> originalDelegates;
-
-    @GuardedBy("lock")
-    final List<Whitelist> delegates = new ArrayList<>();
-
-    @GuardedBy("lock")
-    private final List<EnumeratingWhitelist.MethodSignature> methodSignatures = new ArrayList<>();
-
-    @GuardedBy("lock")
-    private final List<EnumeratingWhitelist.NewSignature> newSignatures = new ArrayList<>();
 
-    @GuardedBy("lock")
-    private final List<EnumeratingWhitelist.MethodSignature> staticMethodSignatures = new ArrayList<>();
-
-    @GuardedBy("lock")
-    private final List<EnumeratingWhitelist.FieldSignature> fieldSignatures = new ArrayList<>();
-
-    @GuardedBy("lock")
-    private final List<EnumeratingWhitelist.FieldSignature> staticFieldSignatures = new ArrayList<>();
-
-    /** anything wrapping us, so that we can propagate {@link #reset} calls up the chain */
-    @GuardedBy("lock")
-    private final Map<ProxyWhitelist,Void> wrappers = new WeakHashMap<>();
-
-    // TODO Consider StampedLock when we switch to Java8 for better performance - https://dzone.com/articles/a-look-at-stampedlock
-    private ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
```

Comment:
FWIW, I think the old code would have blocked `permits*` calls from other threads while `ApprovedWhitelist` was being reconfigured, while the new code with the lock removed reconfigure things on the first thread that sees `ApprovedWhitelist.initialized` changed from `false` to `true`, and other threads will not be blocked and will either use the old or new list depending on timing. Seems ok to me, but I just wanted to note the behavior change.

Reply:
Right, I think the lock was there merely to prevent momentary inconsistencies, not specifically to block `permit*` calls.